### PR TITLE
Add production settings for docker compose

### PR DESCRIPTION
**Pull request by suvash**

### CODE REVIEW

**Comment by suvash** [comment]
@cliftonmcintosh Starting a branch (and PR) for docker for production. Not quite done yet, just cleaned up some deprecated stuff. Also,  a place to talk about all things docker-for-production until it works well.

Need to decide on some stuff before I can add more things. Will have to add one more compose file for production environment. Before that, I have a question for you. Right now, I used the names for environments to be `dev`, `staging` and `prod`, which isn't very consistent. Not sure if that bothers you much, but we could use one of the following
- `dev`, `stage`, `prod`
- `development`, `staging`, `production`
Either works for me, let's decide on one. I'll add the production compose file accordingly. I think it'll be more or less the same as staging, so maybe I can DRY up some stuff there.

File: compose.staging.yml [hunk 1]
```diff
@@ -55,7 +46,7 @@ services:
     environment:
       - APP_ENV
       - CADDYPATH=/etc/caddycerts
-      - PROXY_HOST=web-host
+      - PROXY_HOST=web
       - PROXY_PORT=8000
       - HOST_NAME=staging.nepalmap.org # Change to staging server for nepalmap.org
       - TLS_EMAIL=test@example.com     # Enter a email address here link to LetsEncrypt certs
```
Comment (cliftonmcintosh) [inline compose.staging.yml]: When is the `TLS_EMAIL` variable used?

Is there value in some of these being set via environment variables?

Comment (suvash) [reply compose.staging.yml]: `TLS_EMAIL` is passed on to Caddy ( https://github.com/Code4Nepal/nepalmap_app/blob/dev/compose/caddy/Caddyfile#L3 ). It has to be set if you want to get the certificates non-interactively. https://caddyserver.com/docs/tls

Of course, you can hardcode them in the Caddyfile. But, this way you can override email addresses easily for different environments. The main idea is to have all the configuration injected. ( following ideas from https://12factor.net/config ) It's a matter of opinion really, so feel free to change it if you see fit.

If you don't want to list email, what we can do is pass through the env var like we do for `APP_ENV` which has to be set outside the compose file.

**Comment by suvash** [comment]
@cliftonmcintosh Let me know of the changes above. (I've also dropped a small comment on running Caddy without TLS/cert, for test purposes towards the end of the stage/prod compose file)

File: Makefile [hunk 2]
```diff
@@ -40,7 +55,7 @@ tail-logs: ## Tail the logs for the project service containers (Filtered via SER
 # For deploying
 
 pull-latest:
-	$(if $(APP_ENV_IS_PROD), git checkout master && git pull, $(info -- Not pulling on $(APP_ENV) environment))
+	$(if $(APP_ENV_IS_STAGE_OR_PROD), git checkout master && git pull, $(info -- Not pulling on $(APP_ENV) environment))
```
Comment (cliftonmcintosh) [inline Makefile]: @suvash 

I have typically been deploying the `dev` branch to the test/staging server. Then, once it looks right, I have been merging `dev` into `master` and deploying `master` to production. Any thoughts about whether or not it makes sense to be able to checkout a specific branch here versus always checking out master?

File: compose/caddy/bin/start-caddy [hunk 3]
```diff
@@ -7,6 +7,6 @@ set -eou pipefail
 if [ "$APP_ENV" = "prod" ]; then
     exec /usr/bin/caddy --conf /etc/Caddyfile --log stdout
```
Comment (cliftonmcintosh) [inline compose/caddy/bin/start-caddy]: if the `-ca` tag is not specified, does it default to using the production letsencrypt server?

File: Makefile (hunk 2 above)
Comment (suvash) [reply Makefile]: Ah, sure then. I can get rid of this make target altogether. That way, git branch workflows can be handled outside of the make targets. Keeps things simple as well.

File: compose/caddy/bin/start-caddy (hunk 3 above)
Comment (suvash) [reply compose/caddy/bin/start-caddy]: Yes. I believe so. https://caddyserver.com/docs/automatic-https#testing

**Comment by cliftonmcintosh** [comment]
@suvash 

Thank you for the updates. I will rebuild the staging server this evening with these updates. Assuming that goes well, we can then prepare a new production server in the same way. This will certainly help make our deployment process better.

**Comment by suvash** [comment]
@cliftonmcintosh One thing though. Don't run the `prod` environment for certificate generation before the DNS is properly resolving. If Caddy continuously hits the non-staging Let's Encrypt url and fails(because of DNS), you might be rate limited and not be able to generate certs for a week.

**Comment by cliftonmcintosh** [comment]
@suvash Thanks for the tip about not running 
> the `prod` environment certificate generation before the DNS is properly resolving. 

I have a question. Does Caddy try to reach out to get a certificate for its domain immediately on startup or does it wait for the first request it receives? If it does it immediately on startup and fails, does it try again and again until we get shut out?

I'm just trying to think through the implications for switching the prod server.

**Comment by suvash** [comment]
Yes. Caddy tries to reach out to get a certificate immediately on startup if the tls option is enabled. You can follow the logs to see this. `make tail-logs`.
If the cert generation fails, Caddy will exit. But, since the docker compose file always restart services that crash/exit for some reason, this will start Caddy again (and again)

What you can do is to switch off the TLS temporarily. You can do this by 
- turning off TLS in the Caddyfile. ( `tls off` )
- port mapping 2015 to 80 in compose file. ( `0.0.0.0:80:2015`, because `2015` is the default non-TLS port ) 

 Once you see that the domain name resolves properly, you can revert the changes and use the certs.

**Comment by suvash** [comment]
Also, a note that the docker environment makes use of `requirements.txt` located at `compose/web/` instead of the one located at the project root, and they don't exactly match up.

**Pull request merged**